### PR TITLE
Extend changes for c534 to all nano receivers

### LINKED ISSUE #406: M305 nano receiver 046d:c52f unpairing failed
>Failed to unpair Wireless Mouse M305 from Nano Receiver
>The receiver returned an error, with no further details

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial

$ apt-cache policy solaar
solaar:
  Installed: 0.9.2+dfsg-

### PR DESCRIPTION
**Pull request by pfps**
Make other nano receivers not unpair and replace when pairing.   Extends PR #674

I don't have any such receivers so I am unable to test whether all nano receivers work this way.

Fixes #406

### CODE REVIEW

File: lib/logitech_receiver/receiver.py [hunk 1]
```diff
@@ -334,29 +334,31 @@ def __init__(self, handle, device_info):
 		self.path = device_info.path
 		# USB product id, used for some Nano receivers
 		self.product_id = device_info.product_id
+		product_info = _product_information(self.product_id)
+		if not product_info:
+			raise Exception("unknown receiver type", self.product_id)
```
Comment (FFY00) [inline lib/logitech_receiver/receiver.py]: Capitalize `unknown`.

File: lib/logitech_receiver/receiver.py [hunk 2]
```diff
@@ -334,29 +334,31 @@ def __init__(self, handle, device_info):
...
 
 		# read the serial immediately, so we can find out max_devices
 		# this will tell us if it's a Unifying or Nano receiver
```
Comment (FFY00) [inline lib/logitech_receiver/receiver.py]: Remove the comment please.

File: lib/logitech_receiver/receiver.py [hunk 3]
```diff
@@ -372,6 +374,15 @@ def firmware(self):
 			self._firmware = _hidpp10.get_firmware(self)
 		return self._firmware
 
+	# how many pairings remain (None for unknown, -1 for unlimited)
+	def remaining_pairings(self,cache=True):
+		if self._remaining_pairings is None or not cache :
```
Comment (FFY00) [inline lib/logitech_receiver/receiver.py]: No space before `:`.

File: lib/solaar/ui/window.py [hunk 4]
```diff
@@ -572,6 +572,9 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
 		paired_text += '\n\n<small>%s</small>' % ngettext('Up to %(max_count)s device can be paired to this receiver.', 'Up to %(max_count)s devices can be paired to this receiver.', receiver.max_devices) % { 'max_count': receiver.max_devices }
 	elif(devices_count > 0):
 		paired_text += '\n\n<small>%s</small>' % _('Only one device can be paired to this receiver.')
+	pairings = receiver.remaining_pairings(False)
```
Comment (FFY00) [inline lib/solaar/ui/window.py]: Remove parenthesis please.

File: lib/solaar/cli/unpair.py [hunk 5]
```diff
@@ -27,6 +27,10 @@ def run(receivers, args, find_receiver, find_device):
 	device_name = args.device.lower()
 	dev = find_device(receivers, device_name)
 
+	if not dev.receiver.may_unpair :
```
Comment (FFY00) [inline lib/solaar/cli/unpair.py]: No space before `:`.

File: lib/logitech_receiver/receiver.py (hunk 2 above)
Comment (pfps) [reply lib/logitech_receiver/receiver.py]: You mean the one about reading the serial number?  The second line is incorrect, so I'm removing that.  The first line was there before and is still valid.  Should I remove that as well?

File: lib/logitech_receiver/receiver.py (hunk 3 above)
Comment (pfps) [reply lib/logitech_receiver/receiver.py]: Done.

File: lib/solaar/ui/window.py (hunk 4 above)
Comment (pfps) [reply lib/solaar/ui/window.py]: Done.

File: lib/solaar/cli/unpair.py (hunk 5 above)
Comment (pfps) [reply lib/solaar/cli/unpair.py]: Done.

File: lib/logitech_receiver/receiver.py (hunk 1 above)
Comment (pfps) [reply lib/logitech_receiver/receiver.py]: Done.

File: lib/logitech_receiver/receiver.py (hunk 2 above)
Comment (FFY00) [reply lib/logitech_receiver/receiver.py]: No, that is fine.

File: lib/solaar/ui/window.py [hunk 6]
```diff
@@ -595,8 +595,9 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
 	# b._insecure.set_visible(False)
 	buttons._unpair.set_visible(False)
 
-	may_pair = receiver.may_unpair and not is_pairing
-	if may_pair and devices_count >= receiver.max_devices:
+	may_pair = ( receiver.may_unpair or receiver.re_pairs ) and not is_pairing and \
```
[the remark this answers is not in the document]
Comment (pfps) [reply lib/solaar/ui/window.py]: Done.  I'm not sure how much more readable the new version is.  The logic for determining whether pairing is possible is a bit convoluted.

[the remark this answers is not in the document]
Comment (FFY00) [reply lib/solaar/ui/window.py]: I don't see any update.

File: lib/solaar/ui/window.py [hunk 7]
```diff
@@ -594,12 +594,16 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
...
-	buttons._pair.set_sensitive(may_pair)
+	if ( receiver.may_unpair or receiver.re_pairs ) and not is_pairing and \
+	   ( receiver.remaining_pairings() is None or receiver.remaining_pairings() != 0 ):
+		if not receiver.re_pairs and devices_count >= receiver.max_devices:
+			paired_devices = tuple(n for n in range(1, receiver.max_devices+1) if n in receiver)
+			buttons._pair.set_sensitive(len(paired_devices) < receiver.max_devices)
+		else:
+			buttons._pair.set_sensitive(True)
+	else:
+		buttons._pair.set_sensitive(False)
```
Comment (FFY00) [inline lib/solaar/ui/window.py]: Hum, maybe we can move this to a `can_pair` function?

Comment (pfps) [reply lib/solaar/ui/window.py]: The logic is only needed here so I'm not sure that anything would be gained by moving this elsewhere.

Comment (FFY00) [reply lib/solaar/ui/window.py]: Sure.

**Pull request merged**